controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +220,462 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
 		return ctrl.Result{}, err
 	}
 
-	if err := r.ReconcileNormal(rsCtx); err != nil {
+	result, err := r.ReconcileNormal(rsCtx)
+	if err != nil {
 		rsCtx.Logger.Error(err, "Failed to reconcile VirtualMachineReplicaSet")
 		return ctrl.Result{}, err
 	}
 
-	return ctrl.Result{}, nil
+	return result, nil
+}
+
+func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+	ctx.Logger.Info("Reconciling VirtualMachineReplicaSet Deletion")
+
+	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
+		defer func() {
+			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", nil, false)
+		}()
+
+		controllerutil.RemoveFinalizer(ctx.ReplicaSet, finalizerName)
+	}
+
+	return nil
+}
+
+// shouldExcludeVirtualMachine returns true if the VirtualMachine should be
+// filtered out, false otherwise.
+func shouldExcludeVirtualMachine(rs *vmopv1.VirtualMachineReplicaSet, vm *vmopv1.VirtualMachine) bool {
+	if metav1.GetControllerOf(vm) != nil && !metav1.IsControlledBy(vm, rs) {
+		return true
+	}
+
+	return false
 }
 
-func (r *Reconciler) ReconcileNormal(ctx *pkgctx.VirtualMachineReplicaSetContext) error {
+// adoptOrphan sets the VirtualMachineReplicaSet as a controller OwnerReference
+// to the VirtualMachine.
+func (r *Reconciler) adoptOrphan(
```

Comment:
`adoptOrphan`?

You are going to make the VM feel bad. I mean, what if its ESXi host was *just* lost in a disconnection accident? Too soon @aruneshpa, too soon...